### PR TITLE
Move away from KeyFactoring to a Key builder

### PR DESCRIPTION
**Pull request by gabizou**
Eliminates some implementation detail in `Key`s and how they are used. Allows for some in depth registration to make an indexer for finding keys by plugin as they are registered.

### CODE REVIEW

File: src/main/java/org/spongepowered/api/data/key/KeyFactory.java [hunk 1]
```diff
@@ -71,6 +71,7 @@ private KeyFactory() {}
      * @param name The name for the new key
      * @return The generated key
      */
+    @Deprecated
```
Comment (Deamon5550) [inline src/main/java/org/spongepowered/api/data/key/KeyFactory.java]: Need some deprecation javadocs here with the alternative

**Review by Deamon5550 — APPROVED** [review]
lgtm pending javadocs

**Comment by JBYoshi** [comment]
Will the deprecated `KeyFactory` methods be rewritten to delegate to `Key.Builder`?

**Comment by liach** [comment]
I believe the whole class is going to be removed.

File: src/main/java/org/spongepowered/api/data/key/Key.java [hunk 2]
```diff
@@ -66,4 +74,23 @@
      */
     DataQuery getQuery();
 
+    interface Builder<E, V extends BaseValue<E>> extends ResettableBuilder<Key<V>, Builder<E, V>> {
+
+        <T, B extends BaseValue<T>> Builder<T, B> valueElement(TypeToken<T> element, TypeToken<B> token);
```
Comment (Cybermaxke) [inline src/main/java/org/spongepowered/api/data/key/Key.java]: Just the value typetoken is sufficient, you can extract the element typetoken from the value typetoken.

Comment (Cybermaxke) [reply src/main/java/org/spongepowered/api/data/key/Key.java]: I would also name the method differently, maybe `valueType`?

File: src/main/java/org/spongepowered/api/data/key/Key.java [hunk 3]
```diff
@@ -66,4 +74,23 @@
...
+
+        Builder<E, V> keyId(String id);
```
Comment (Cybermaxke) [inline src/main/java/org/spongepowered/api/data/key/Key.java]: Just `id`?

Comment (Cybermaxke) [reply src/main/java/org/spongepowered/api/data/key/Key.java]: We should be more consistent with builder methods for catalog types:
https://github.com/SpongePowered/SpongeAPI/issues/1721

**Pull request merged**